test/units/modules/cloud/docker/test_docker_swarm_service.py

```diff
@@ -169,60 +170,89 @@ def test_has_dict_changed(docker_swarm_service):
 
 
 def test_has_list_changed(docker_swarm_service):
+
+    # List comparisons without dictionaries
+    # I could improve the indenting, but pycodestyle wants this instead
+    assert not docker_swarm_service.has_list_changed(None, None)
+    assert not docker_swarm_service.has_list_changed(None, [])
+    assert not docker_swarm_service.has_list_changed(None, [1, 2])
+
+    assert not docker_swarm_service.has_list_changed([], None)
+    assert not docker_swarm_service.has_list_changed([], [])
+    assert docker_swarm_service.has_list_changed([], [1, 2])
+
+    assert docker_swarm_service.has_list_changed([1, 2], None)
+    assert docker_swarm_service.has_list_changed([1, 2], [])
+
+    assert docker_swarm_service.has_list_changed([1, 2, 3], [1, 2])
+    assert docker_swarm_service.has_list_changed([1, 2], [1, 2, 3])
+
+    # Check list sorting
+    assert not docker_swarm_service.has_list_changed([1, 2], [2, 1])
     assert docker_swarm_service.has_list_changed(
-        [
-            {"a": 1},
-            {"b": 1}
-        ],
-        [
-            {"a": 1}
-        ]
+        [1, 2],
+        [2, 1],
+        sort_lists=False
     )
+
+    # Check type matching
+    assert docker_swarm_service.has_list_changed([None, 1], [2, 1])
+    assert docker_swarm_service.has_list_changed([2, 1], [None, 1])
     assert docker_swarm_service.has_list_changed(
-        [
-            {"a": 1},
-        ],
-        [
-            {"a": 1},
-            {"b": 1},
-        ]
-    )
-    assert not docker_swarm_service.has_list_changed(
-        [
-            {"a": 1},
-            {"b": 1},
-        ],
-        [
-            {"a": 1},
-            {"b": 1}
-        ]
-    )
-    assert not docker_swarm_service.has_list_changed(
-        None,
-        [
-            {"b": 1},
-            {"a": 1}
-        ]
+        "command --with args",
+        ['command', '--with', 'args']
     )
     assert docker_swarm_service.has_list_changed(
-        [],
-        [
-            {"b": 1},
-            {"a": 1}
-        ]
+        ['sleep', '3400'],
+        [u'sleep', u'3600'],
+        sort_lists=False
     )
+
+    # List comparisons with dictionaries
     assert not docker_swarm_service.has_list_changed(
-        None,
-        None
+        [{'a': 1}],
+        [{'a': 1}],
+        sort_key='a'
     )
+
     assert not docker_swarm_service.has_list_changed(
-        [],
-        None
+        [{'a': 1}, {'a': 2}],
+        [{'a': 1}, {'a': 2}],
+        sort_key='a'
     )
+
+    if python_version_info[0] >= 3:
+        # This test does not work on Python 2
```

Comment:
Maybe it's better to modify `has_list_changed` so it gives the same results for Python 2 **and** Python 3?

Reply:
I've updated the code to make this consistent. The exception should only be thrown on a programming error.